README
NAME
    DBIx::Inline - DBIx::Class without the class.

DESCRIPTION
    This module is yet another interface to DBI. I like how DBIx::Class
    works, separating the results from the resultsets, the resultsets from
    the results and the schema from everything else. It's tidy, easy to
    follow and works a treat. I also like how you can "reuse" queries in
    resultsets and results without typing them out again and again. However,
    when I wanted to work on a small project I found DBIx::Class a little
    slow and didn't want to keep setting up the classes for it to work.
    DBIx::Inline attempts follow the way DBIx::Class does things, but more
    "inline". You still get the reusable queries, Results and ResultSets,
    but without all the classes to setup. You do lose a lot of functionality
    that you get with DBIx::Class, but that's not what DBIx::Inline is
    really about. I wanted it to be faster and not hold your hand with
    everything, yet still be easy enough to use. It's still possible to have
    accessors and Result/ResulSet methods, but they are created on-the-fly
    with method. Also, you can automatically create all accessors for a
    result using load_accessors. DBIx::Inline is great for small projects
    that do not require a lot of customisation, but for anything else I'd
    highly recommend DBIx::Class.

SYNOPSIS
        package MyDB;

        use base 'DBIx::Inline';

        my $rs = MyDB->model('Foo')->all; # Read up about models to see what this does
        # or..
        my $rs = MyDB->sqlite('/some/sqlite.db')->resultset('users')->all;
        # or..
        my $rs = MyDB->connect(
            dbi => 'SQLite:/some/sqlite.db',
        );
        $rs = $rs->resultset('users');
    
        # create a resultset method on-the-fly
        $rs->method(not_active => sub {
            return shift->search([], { account_status => 'disabled' }, { order => ['id'], rows => 5 });
        });

        # chain the custom resultset method with a core one (count)
        print "Rows returned: " . $rs->not_active->count . "\n";

        # make the records in the resultset active
        # will return a resultset with the updated data
        my $new_rs = $rs->update({account_status => 'active'});

  connect
    Creates the Schema instance using the hash specified. Currently only dbi
    is mandatory, which tells DBI which engine to use (SQLite, Pg, etc). If
    you're using SQLite there is no need to set user or pass.

        my $dbh = DBIx::Inline->connect(
            dbi => 'SQLite:/var/db/test.db',
        );

        my $dbh = DBIx::Inline->connect(
            dbi  => 'Pg:host=myhost;dbname=dbname',
            user => 'username',
            pass => 'password',
        );

  model
    Models make your life easier when you need to reuse a specific
    connection. You can even go so far as specifying a ResultSet to use by
    default. By default, DBIx::Inline will look for inline.yml, unless you
    have configured a different models file with "config('file.yml')". The
    syntax is very basic and uses a simple YAML file, making it easy to move
    around if you need to.

        # inline.yml
        ---
        Foo:
          connect: 'SQLite:foo.db'
    
        AnotherSchema:
          connect: 'Pg:host=localhost;dbname=foo'
          user: 'myuser'
          pass: 'pass'

        WithResultSet:
          connect: 'SQLite:test.db'
          table: 'users'
          columns: id name status date_created(datetime)
 
        # test.pl
        package main;
  
        my $rs = main->model('AnotherSchema')->resultset('the_table');
        my $rs2 = main->model('WithResultset'); # that's all we need!
        while(my $row = $rs2->next) {
            $row->load_accessors;
            print $row->name;
        }

    As of 0.15 you can now use related tables. It basically does a
    search_join in a convenient accessor for you. The accessor search is
    *very* limited, allowing only one key.

        # inline.yml
        AnotherSchema:
          connect: 'Pg:host=localhost;dbname=foo'
          user: 'myuser'
          pass: 'pass'
          related:
            authors: 'id <-> books(authors_id)'

        # then in your code
        my $rs = $c->model('AnotherSchema')->resultset('authors');
        my $books = $rs->authors({ id => 3 }); # search for all books by author with id of 3

        # now use it as any normal resultset
        while( my $row = $books->next ) {
            $row->load_accessors;
            print $row->book_title;
        }

    As of 0.17 you can now resuse a models connection.

        ---
        MainDB:
          connect: 'Pg:host=localhost;dbname=thisdb'
          user: 'foo'
          pass: 'bar'
    
        Author:
          model: MainDB
          resultset: authors
          related:
            authors: 'id <-> books(author_id)'
   
        BooK:
          model: MainDB
          table: books

    0.19 offers the ability to add columns. Doing this will create an
    accessor to that result. You can even return a DateTime object on time
    results by adding "(datetime)" at the end of the column.

        ---
        Customer:
          model: MyModel
          table: customers
          columns: id name status date_created(datetime) suspended(datetime)

        # sql.pl
        package MyDB;
 
        my $customers = MyDB->model('Customer')->all;
        print $customers->last->date_created;      # prints DateTime object
        print $customers->last->date_created->ymd; # you can then call DateTime methods on the object

  sqlite
    Initially load a SQLite database (file). Instead of going through models
    or dbi string we can just call "sqlite('file')".

        package main;
    
        use base 'DBIx::Inline';

        my $schema = main->sqlite('/path/to/db.db')->resultset('users');

  config
    Sets the location of the configuration (file with the models. The Schema
    models.. not girls). This allows you to have the file anywhere on your
    system and you can rename it to anything.

        # /var/schema/myschemas.yml
        Foo:
          connect: 'SQLite:/var/db/mydb.db'
    
        # /scripts/db.pl
        package main;

        use base 'DBIx::Inline';

        main->config ('/var/schema/myschemas.yml');
        my $schema = main->model('Foo');

    You can even chain "config" to "model" if you want.

        my $schema = main->config('/var/schema/myschemas.yml')->model('Foo');

AUTHOR
    Brad Haywood <brad@geeksware.net>

LICENSE
    You may distribute this code under the same terms as Perl itself.



Changes
0.19        13/Dec/2011
            You can specify 'columns' in the models file now. Any columns listed in there will have accessors set up for them in the Result class.
            Now, you can even add a DateTime attribute to a specific column or columns. When you do this the result will be in the form of a DateTime object, if possible.

0.18        06/Dec/2011
            Fixed problem where you could not have a relationships and resultsets in the same model file.
            Added option 'model' to config file to "reuse" a models connection inside the same config.

0.16        02/Dec/2011
            Added 'related' option to config file.

0.15        02/Dec/2011
            Altered 'model' to let you add a resultset to the models file. Just use 'resultset: YourTableName' and it will automatically load it for you.
            This will also be in the documentation, too.

0.14        02/Dec/2011
            Hopefully fixed broken test. Changed the test module to Test::Lite.
            I honestly hope this doesn't screw anyone over, but table is now, again, resultset (Won't happen again. Promise).

0.13        22/Nov/2011
            Added the 'search_join' resultset method. Check out L<DBIx::Inline::ResultSet> for more information

0.12        22/Nov/2011
            Sorry about the major change, but the Schema method 'resultset' has been changed to 'table'.
            I do apologize, but I couldn't live with it anymore.. it was a TABLE not a ResultSet.. argh.

0.11        22/Nov/2011
            Added sqlite method to main class. Now you can open a SQLite file 
            without having to go through a model or dbi strings.


lib/DBIx/Inline/Manual.pm
package DBIx::Inline::Manual;

=head1 NAME

DBIx::Inline::Manual - A manual for DBIx::Inline

=head1 DESCRIPTION

The users manual for L<DBIx::Inline>. The pod for the actual module was beginning to get too big, so I decided to move it to its 
own manual.

=head1 CONTENT

=head2 Starting off

B<Connecting>

There are two ways to begin using DBIx::Inline - using models, or connecting directly within your package.
If you're wanting to reuse the same schema over and over again you're best method would to create a file called inline.yml, put your 
configuration in there and reuse the model when you need it. If it's just going to be a small once-off, it may be easier to just 
connect directly.

B<Using Models>

A simple inline.yml file will look like this

  ---
  MySchema:
    connect: 'SQLite:mydb.db'

  FooSchema:
    connect: 'Pg:host=localhost;dbname=foo'
    user: 'foo',
    pass: 'foopass'

That gives us two reusable schemas. One using a simple SQLite driver, and the other Postgres. To use them in our package we reference them 
using B<model> like so.

  package main;
 
  my $schema = main->model('MySchema');
  my $foo = main->model('FooSchema')->resultset('a_table')->all;

We successfully now have two schemas running in the same package. The first one returns a schema, the second one chained a resultset onto it.
Once your schema has a resultset (at this point a resultset is just a table.. I need to rename the schema method from resultset to table, really), you 
can start searching, finding, updating, etc. A resultset is just a class holding multiple records.

=head2 ResultSet

B<Searching>

The main part to doing anything at all in DBIx::Inline is returning a resulset via search. Search is a powerful method with a few different options. I'll show you 
how to return a simple resultset, and how to use it to paginate results for a web query with ease.
First up, let's perform a simple search.

  # search the table where the rows status = 'active', but only return the id and name
  my $rs = $table->search([qw/id name/], { status => 'active' });

  # search the table where the rows status = 'active', but only return 5 results
  # also, order by id first
  my $rs = $table->search([], { status => 'active' }, { rows => 5, order => ['id'] });
  
  # now paginate the entire resultset returning 5 records at a time
  my $page1 = $table->all->search([], {}, { page => 1, rows => 5 });
  my $page2 = $table->all->search([], {}, { page => 2, rows => 5 });

  # get the first and last records of a result by chaining
  my $first = $table->all->first;
  my $last = $table->search([], { code => 5485 })->last;

B<What to do with ResultSet results>

You can iterate through results using B<next>.

  while(my $row = $rs->next) {
      say $row->{column_name};
  }

Alternatively, you can update or delete them

  my $rs = $table->search([], { foo => 'baz' });
  $rs->update({ status => 'active' });
  $rs->delete;

Less SQL queries = happy.

When you iterate through a resultset using next, it returns the result as a L<DBIx::Inline::Result>.
Another way to return a result from a resultset is using B<find>. If you use 'find' it will return a single 
row.

  my $result = $rs->find([], { foo => 'baz' });
  say $result->{name};

As you can see it works the same as search, but will not return a resultset, and only returns 1 row.

=head2 Results

There isn't a great deal you can do with results compared to resulsets - they simply hold the row data for the current 
result in a hash.

  $result->{name}
  $result->{id}
  etc..

You can create accessors for all of columns with B<load_accessors>

  $result->load_accessors;
  say $result->name;
  say $result->id;

But what if you have long column names and want them shortened? No problem, use B<accessorize>.

  $result->accessorize(
      name => 'long_name_column',
      id   => 'repeat_id',
  );

  say $result->name; # instead of $result->{long_name_column}

=cut


lib/DBIx/Inline/Result.pm
package DBIx::Inline::Result;

=head1 NAME

DBIx::Inline::Result - Class for DBIx::Inline results

=cut 

use Goose ':Antlers';
use SQL::Abstract::More;
our $sql = SQL::Abstract::More->new;

use vars qw/$sql/;

our $VERSION = '0.14';

sub method {
    my ($self, %args) = @_;

    my $key;
    for (keys %args) {
        $key = $_;
    }
    *$key = sub { $args{$key}->($self); };

    bless \*$key, 'DBIx::Inline::Result';
}

=head2 load_accessors

Creates all accessors for the current result. So instead of using hashes you can use 
subroutines.

    my $row = $rs->find([], { id => 2 });

    $row->load_accessors;
    print $row->name . "\n";
    print $row->id . "\n";

These accessors can also be used to set new values just by adding arguments.

    $row->load_accessors;
    $row->notes('Updated notes');
    $row->name('My New Name');

=cut

sub load_accessors {
    my $self = shift;
    for (keys %$self) {
        my $result = $self->{$_};
        my $table = $self->{_from};
        my $where = $self->{_where};
        my $schema = $self->{_schema};
        my $key = $_;
        *$_ = sub {
            my ($self, $a) = @_;
            if (! $a) { return $result; }
            else {
                my ($stmt, @bind) = $sql->update($table, { $key => $a }, $where);
                my $sth = $schema->prepare($stmt);
                $sth->execute(@bind);
            }
        };
        bless \*$_, 'DBIx::Inline::Result';
    }
}

=head2 accessorize

The same as load_accessors, but will only create the ones you specify. Also, it allows you 
to name the accessor to whatever you want.

    $row->accessorize(
        name    => 'long_winded_column_name',
        status  => 'user_status'
    );

    $row->name; # will fetch $row->{long_winded_column_name};

=cut

sub accessorize {
    my ($self, %args) = @_;

    for (keys %args) {
        my $result = $self->{$args{$_}};
        *$_ = sub { return $result; };
        bless \*$_, 'DBIx::Inline::Result';
    }
}

1;


lib/DBIx/Inline/ResultSet.pm
package DBIx::Inline::ResultSet;

=head1 NAME

DBIx::Inline::ResultSet - Methods for searching and altering tables

=cut
use Goose;
use SQL::Abstract::More;

our $sql = SQL::Abstract::More->new;
use vars qw/$sql/;

our $VERSION = '0.14';

=head2 fetch

Fetches the results for the entire resulset

=cut

sub fetch {
    my $self = shift;

    return $self->{result}
}

=head2 count

Returns the number of rows found

=cut

sub count {
    my $self = shift;

    return scalar @{$self->{result}};
}

=head2 all

Returns all rows in a table as a resultset

    my $rs = $schema->resultset('Users')->all;

=cut

sub all {
    my $self = shift;

    return $self->search([], {});
}

=head2 first

Get the first result from a resultset

=cut

sub first {
    my $self = shift;

    return bless $self->{result}->[0], 'DBIx::Inline::Result';
}

=head2 last

Get the last result from a resultset

=cut

sub last {
    my $self = shift;

    return bless $self->{result}->[ scalar(@{$self->{result}}-1) ], 'DBIx::Inline::Result';
}

=head2 next

A simple iterator to loop through a resultset. Each 
returned result will be returned as a Result.

    while(my $row = $result->next) {
        print $row->{name};
    }

=cut

sub next {
    my $self = shift;
    if (! exists $self->{_it_pos}) {
        $self->{_it_pos} = 0;
        $self->{_it_max} = scalar @{$self->{result}};
    }
    my $pos = $self->{_it_pos};
    $self->{_it_pos}++;
    if ($self->{_it_pos} > $self->{_it_max}) {
        delete $self->{_it_pos};
        delete $self->{_it_max};
        return undef;
    }
    return bless $self->{result}->[$pos], 'DBIx::Inline::Result';
}

=head2 primary_key

Sets the primary key for the current ResultSet

    $rs->primary_key('id');

=cut

sub primary_key {
    my ($self, $key) = @_;

    return 0 if ! $key;
    
    $self->{primary_key} = $key;
    return $self;
}

=head2 search

Access to the SQL SELECT query. Returns an array with the selected rows, which contains a hashref of values.
First parameter is an array of what you want returned ie: SELECT this, that
If you enter an empty array ([]), then it will return everything ie: SELECT *
The second parameter is a hash of keys and values of what to search for.

    my $res = $resultset->search([qw/name id status/], { status => 'active' });

    my $res = $resultset->search([], { status => 'disabled' });
    
    my $res = $resultset->search([], { -or => [ name => 'Test', name => 'Foo' ], status => 'active' });

    my $res = $resultset->search([],
        { status => 'active' },
        {
            order => ['id'],
            rows => 10,
        }
    );

=cut

sub search {
    my ($self, $fields, $c, $order) = @_;
    if (scalar @$fields == 0) { push @$fields, '*'; }
    if (exists $self->{where}) {
        for (keys %{$self->{where}}) {
            $c->{$_} = $self->{where}->{$_};
        }
    }

    $order->{rows}
        if exists $self->{rows};
    $order->{page}
        if exists $self->{page};

    my %args;
    $args{-columns} = $fields;
    $args{-from} = $self->{table};
    $args{-where} = $c;
    $args{-order_by} = $order->{order} if exists $order->{order};
    if (exists $order->{rows}) {
        if (exists $order->{page}) {
            $args{-page_size} = $order->{rows};
        }
        else { $args{-limit} = $order->{rows}; }
    }
    $args{-page_index} = $order->{page} if exists $order->{page};
    my ($stmt, @bind) = $sql->select(
        %args,
    );
    #my ($stmt, @bind) = $sql->select($self->{table}, $fields, $c);
    my ($wstmt, @wbind) = $sql->where($c);
        
    my $result = {
        dbh    => $self->{dbh},
        result => $self->{dbh}->selectall_arrayref($stmt, { Slice => {} }, @bind),
        stmt   => $wstmt,
        bind   => \@wbind,
        where  => $c,
        table  => $self->{table},
        primary_key => $self->{primary_key},
        r           => 'DBIx::Inline::Result',
        rs          => __PACKAGE__,
    };
    $result->{rows} = $order->{rows}
        if exists $order->{rows};
    
    $result->{page} = $order->{page}
        if exists $order->{page};

    return bless $result, __PACKAGE__;
}

sub find {
    my ($self, $fields, $c) = @_;
    if (scalar @$fields == 0) { push @$fields, '*'; }
    my ($stmt, @bind) = $sql->select(
        -columns => $fields,
        -from    => $self->{table},
        -where   => $c,
        -limit   => 1
    );
    my ($wstmt, @wbind) = $sql->where($c);
    my $result = $self->{dbh}->selectall_arrayref($stmt, { Slice => {} }, @bind)->[0];
    $result->{_where} = $c;
    $result->{_from} = $self->{table};
    $result->{_schema} = $self->schema;
    return bless $result, 'DBIx::Inline::Result';
}

=head2 insert

Inserts a new record.

    my $user = $resultset->insert({name => 'Foo', user => 'foo_bar', pass => 'baz'});
    $user->load_accessors;
    print $user->name . "\n";
    print "Last inserted primary key: " . $resultset->insert_id . "\n";
=cut

sub insert {
    my ($self, $c) = @_;
   
    die "Cannot insert without primary key\n"
        if ! exists $self->{primary_key};
 
    my ($stmt, @bind) = $sql->insert($self->{table}, $c);
    my $sth = $self->{dbh}->prepare($stmt);
    my $result = $sth->execute(@bind);

    my $res = $self->search([], $c);
    $c->{id} = $self->insert_id;
    ($stmt, @bind) = $sql->select($self->{table}, ['*'], $c);
    
    if ($res->count) {
        my $result = $self->{dbh}->selectall_arrayref($stmt, { Slice => {} }, @bind)->[0];
        $result->{_where} = $c;
        $result->{_from} = $self->{table};
        $result->{_schema} = $self->schema;
        
        return bless $result, 'DBIx::Inline::Result';
    }
    else { return 0; }    
}

=head2 update

Updates the current result using the hash specified

    my $res = $dbh->resultset('foo_table')->search([], { id => 5132 });
    if ($res->update({name => 'New Name'})) {
        print "Updated!\n";
    }

=cut

sub update {
    my ($self, $fieldvals) = @_;

    my ($stmt, @bind) = $sql->update($self->{table}, $fieldvals, $self->{where});
    my $sth = $self->{dbh}->prepare($stmt);
    my %args;
    $args{-columns} = ['*'],
    $args{-from}    = $self->{table},
    $args{-where}   = $self->{where},
    $args{-limit}   = $self->{rows}
        if exists $self->{rows};

    if ($sth->execute(@bind)) {
        my ($sstmt, @sbind) = $sql->select(
            %args,
        );
        my $rs = {
            dbh    => $self->{dbh},
            where  => $self->{where},
            table  => $self->{table},
            result => $self->{dbh}->selectall_arrayref($sstmt, { Slice => {} }, @sbind),
            primary_key => $self->{primary_key},
            r       => 'DBIx::Inline::Result',
            rs      => __PACKAGE__,
        };
        return bless $rs, __PACKAGE__;
    }
    else { return 0; }
}

=head2 insert_id

Gets the primary key value of the last inserted row. It will require the primary key to be set, though

=cut

sub insert_id {
    my ($self) = @_;

    if (! exists $self->{primary_key}) {
        warn "Can't call insert_id on result when no primary_key was defined in the ResultSet";
        return 0;
    }
    if (exists $self->{result}->[scalar(@{$self->{result}})-1]->{$self->{primary_key}}) {
        return $self->{result}->[scalar(@{$self->{result}})-1]->{$self->{primary_key}};
    }

    return 0;
}

=head2 delete

Drops the records in the current search result

    my $res = $resultset->search([], { id => 2 });
    $res->delete; # gone!

=cut

sub delete {
    my ($self) = @_;

    my ($stmt, @bind) = $sql->delete(-from => $self->{table}, -where => $self->{where});

    my $sth = $self->{dbh}->prepare($stmt);
    $sth->execute(@bind);
}

=head2 method

Creates a ResultSet method on the fly. Use it to create accessors, or shortcuts

    $rs->method(is_active => sub {
        return shift->search([], { account_status => 'active' });
    });

    print "Rows: " . $rs->is_active->count . "\n";

=cut

sub method {
    my ($self, $name, $sub) = @_;
    *$name = sub { $sub->(@_) };
    *{"DBIx::Inline::ResultSet::$name"} = \*$name;
}

sub schema {
    my $self = shift;

    return $self->{dbh};
}

=head2 search_join

Performs a SQL JOIN to fetch "related" records.
Say you want to find what books belong to what author. We want to find books matching author ID 2.

    my $authors = $dbi->model('Authors')->table('authors');
    $authors->search_join([], { authors => 'books', on => [ qw(id author_id) ] }, { id => 2 });

=cut

sub search_join {
    my ($self, $fields, $args, $where) = @_;

    my ($table1,$table2,$on);
    if (scalar @$fields == 0) { $fields = '*'; }
    else { $fields = join q{,}, @$fields; }
    
    for (keys %$args) {
        unless ($_ eq 'on') {
            $table1 = $_;
            $table2 = $args->{$_};
        }
        
        if ($_ eq 'on') {
            $on = $args->{$_};
        }
    }

    my ($w_key, $w_val);
    if ($where) {
        if (scalar keys %$where > 0) {
            for (keys %$where) {
                $w_key = $_;
                $w_val = $where->{$_};
            }
        }
    }
    
    my $stmt = "SELECT $fields FROM $table1 AS a INNER JOIN $table2 AS b ON ( a.$on->[0] = b.$on->[1] )";
    if (scalar keys %$where > 0) {
        $stmt .= " WHERE a.$w_key = $w_val";
        $stmt .= "AND (" . $self->{where} . ")"
            if exists $self->{where};
    }
    
    my $result = {
        dbh    => $self->{dbh},
        result => $self->{dbh}->selectall_arrayref($stmt, { Slice => {} }),
        table  => $self->{table},
        primary_key => $self->{primary_key},
        r       => 'DBIx::Inline::Result',
        rs      => __PACKAGE__,
    };
    return bless $result, 'DBIx::Inline::ResultSet'; 
}

1;


lib/DBIx/Inline.pm
package DBIx::Inline;

use DateTime;
use Goose ':Antlers'; 
use DBI;

has 'columns' => ( is => 'rw', default => ['*'] );

extends qw/
    DBIx::Inline::Schema
    DBIx::Inline::ResultSet
    DBIx::Inline::Result
/;

$DBIx::Inline::VERSION = '0.19';
our $global = {};

=head1 NAME

DBIx::Inline - DBIx::Class without the class.

=head1 DESCRIPTION

This module is yet another interface to DBI. I like how L<DBIx::Class> works, separating the results from the resultsets, the resultsets from the results and the schema from everything else. 
It's tidy, easy to follow and works a treat. I also like how you can "reuse" queries in resultsets and results without typing them out again and again. However, when I wanted to work on a small 
project I found DBIx::Class a little slow and didn't want to keep setting up the classes for it to work. DBIx::Inline attempts follow the way DBIx::Class does things, but more "inline". You 
still get the reusable queries, Results and ResultSets, but without all the classes to setup. You do lose a lot of functionality that you get with DBIx::Class, but that's not what DBIx::Inline is 
really about. I wanted it to be faster and not hold your hand with everything, yet still be easy enough to use. 
It's still possible to have accessors and Result/ResulSet methods, but they are created on-the-fly with B<method>. Also, you can automatically create all accessors for a result using B<load_accessors>.
DBIx::Inline is great for small projects that do not require a lot of customisation, but for anything else I'd highly recommend B<DBIx::Class>.

=head1 SYNOPSIS

    package MyDB;

    use base 'DBIx::Inline';

    my $rs = MyDB->model('Foo')->all; # Read up about models to see what this does
    # or..
    my $rs = MyDB->sqlite('/some/sqlite.db')->resultset('users')->all;
    # or..
    my $rs = MyDB->connect(
        dbi => 'SQLite:/some/sqlite.db',
    );
    $rs = $rs->resultset('users');
    
    # create a resultset method on-the-fly
    $rs->method(not_active => sub {
        return shift->search([], { account_status => 'disabled' }, { order => ['id'], rows => 5 });
    });

    # chain the custom resultset method with a core one (count)
    print "Rows returned: " . $rs->not_active->count . "\n";

    # make the records in the resultset active
    # will return a resultset with the updated data
    my $new_rs = $rs->update({account_status => 'active'});

=cut

=head2 connect

Creates the Schema instance using the hash specified. Currently only dbi is mandatory, 
which tells DBI which engine to use (SQLite, Pg, etc).
If you're using SQLite there is no need to set user or pass.

    my $dbh = DBIx::Inline->connect(
        dbi => 'SQLite:/var/db/test.db',
    );

    my $dbh = DBIx::Inline->connect(
        dbi  => 'Pg:host=myhost;dbname=dbname',
        user => 'username',
        pass => 'password',
    );

=cut

sub connect {
    use vars qw/$global/;
    my ($class, %args) = @_;

    my $dbh = DBI->connect(
        'dbi:' . $args{dbi},
        $args{user}||undef,
        $args{pass}||undef,
        { PrintError => 0 }
    ) or do {
        warn 'Could not connect to database: ' . $DBI::errstr;
        return 0;
    };

    my $dbhx = { dbh => $dbh, schema => $class };
    bless $dbhx, 'DBIx::Inline::Schema';
}

=head2 model

Models make your life easier when you need to reuse a specific connection. You can even go so far as specifying a ResultSet to use by default.
By default, DBIx::Inline will look for inline.yml, unless you have configured a different models file with C<config('file.yml')>.
The syntax is very basic and uses a simple YAML file, making it easy to move around if you need to.

    # inline.yml
    ---
    Foo:
      connect: 'SQLite:foo.db'
    
    AnotherSchema:
      connect: 'Pg:host=localhost;dbname=foo'
      user: 'myuser'
      pass: 'pass'

    WithResultSet:
      connect: 'SQLite:test.db'
      table: 'users'
      columns: id name status date_created(datetime)
 
    # test.pl
    package main;
  
    my $rs = main->model('AnotherSchema')->resultset('the_table');
    my $rs2 = main->model('WithResultset'); # that's all we need!
    while(my $row = $rs2->next) {
        $row->load_accessors;
        print $row->name;
    }

As of 0.15 you can now use related tables. It basically does a search_join in a convenient accessor for you. The accessor search is *very* limited, allowing only one key.

    # inline.yml
    AnotherSchema:
      connect: 'Pg:host=localhost;dbname=foo'
      user: 'myuser'
      pass: 'pass'
      related:
        authors: 'id <-> books(authors_id)'

    # then in your code
    my $rs = $c->model('AnotherSchema')->resultset('authors');
    my $books = $rs->authors({ id => 3 }); # search for all books by author with id of 3

    # now use it as any normal resultset
    while( my $row = $books->next ) {
        $row->load_accessors;
        print $row->book_title;
    }

As of 0.17 you can now resuse a models connection.

    ---
    MainDB:
      connect: 'Pg:host=localhost;dbname=thisdb'
      user: 'foo'
      pass: 'bar'
    
    Author:
      model: MainDB
      resultset: authors
      related:
        authors: 'id <-> books(author_id)'
   
    BooK:
      model: MainDB
      table: books

0.19 offers the ability to add columns. Doing this will create an accessor to that result. You can even return a DateTime object on time results by adding C<(datetime)> at the end of the column.

    ---
    Customer:
      model: MyModel
      table: customers
      columns: id name status date_created(datetime) suspended(datetime)

    # sql.pl
    package MyDB;
 
    my $customers = MyDB->model('Customer')->all;
    print $customers->last->date_created;      # prints DateTime object
    print $customers->last->date_created->ymd; # you can then call DateTime methods on the object

=cut

sub model {
    use vars qw/$global/;
    my ($class, $model) = @_;
   
    my $file = $global->{config}||'inline.yml'; 
    die "Can't locate config '$file'\n"
        if ! -f $file;
    
    use YAML::Syck;

    my $yaml = LoadFile($file);
    die "No such model '$model'\n"
        if ! exists $yaml->{$model};
    
    my $dbh;
    
    if (exists $yaml->{$model}->{model}) {
        my $use_model = $yaml->{$model}->{model};
        $dbh = DBI->connect(
            'dbi:' . $yaml->{$use_model}->{connect},
            $yaml->{$use_model}->{user}||undef,
            $yaml->{$use_model}->{pass}||undef,
        );
    }
    else {
        $dbh = DBI->connect(
            'dbi:' . $yaml->{$model}->{connect},
            $yaml->{$model}->{user}||undef,
            $yaml->{$model}->{pass}||undef,
        );
    }

    my $dbhx = { dbh => $dbh, schema => $class };
    
    if (exists $yaml->{$model}->{related}) {
        if (scalar keys %{$yaml->{$model}->{related}} > 0) {
            foreach my $relate (keys %{$yaml->{$model}->{related}}) {
                my $relation = $yaml->{$model}->{related}->{$relate};
                if ($relation =~ /^(\w+)\s*?<\->\s*?(\w+)\s*?\(\s*?(\w+)\s*?\)$/) {
                    my $relate_table = $2;
                    my $relation_id = $1;
                    my $related_id = $3;
                    DBIx::Inline::ResultSet->create( $relate => sub {
                        my ($self, $args) = @_;
                        $self->{table} = $relate;
                        $self->{where} = $args;
                        return $self->search_join([], { $relate => $relate_table, on => [$relation_id, $related_id] }, $args||{} );
                    });
                }
            }
        }
    }

    if (exists $yaml->{$model}->{columns}) {
        my @cols = split ' ', $yaml->{$model}->{columns};
        my $datetime = 0;
        for my $c (@cols) {
            if ($c =~ /(.+)\(datetime\)/) { $datetime = 1; $c = $1; }
            if ($datetime) {
                DBIx::Inline::Result->create( $c => sub {
                    my ($self, $a) = @_;
                    my $result = $self->{$c};
                    my $sql = new SQL::Abstract::More;
                    my $where = $self->{_where};
                    my ($year, $month, $day, $hour, $min, $sec, $dt);
                    if ($result =~ /^(.+)\-(.+)\-(.+) (.+):(.+):(.+)/) {
                        ($year, $month, $day, $hour, $min, $sec) = ($1, $2, $3, $4, $5, $6); 
                        $dt = DateTime->new( day => $day, month => $month, year => $year, hour => $hour, minute => $min, second => $sec );
                    }
                    elsif ($result =~ /^(.+)\-(.+)\-(.+)$/) {
                        ($year, $month, $day) = ($1, $2, $3);
                        $dt = DateTime->new( day => $day, month => $month, year => $year );
                    }
                    
                    return $dt;
                });
            }
            else {
                DBIx::Inline::Result->create( $c => sub {
                    my ($self, $a) = @_;
                    my $result = $self->{$c};
                    my $sql = new SQL::Abstract::More;
                    my $where = $self->{_where};
                    if (! $a) {
                        return $result;
                    }
                    else {
                        my ($stmt, @bind) = $sql->update($table, { $key => $a }, $where);
                        my $sth = $self->_dbh->prepare($stmt);
                        $sth->execute(@bind);
                        $self->{$key} = $a;
                        return $self;
                    }
                });
            }
            $datetime = 0;
        }
    }
   
    if (exists $yaml->{$model}->{table}) {
        
        bless $dbhx, 'DBIx::Inline::Schema';
        my $rs = $yaml->{$model}->{table};
        # get columns from model file?
        
        return $dbhx->resultset($rs)->search([], {});
    }

    bless $dbhx, 'DBIx::Inline::Schema';
}

=head2 sqlite

Initially load a SQLite database (file). Instead of going through models or dbi string we can just call C<sqlite('file')>.

    package main;
    
    use base 'DBIx::Inline';

    my $schema = main->sqlite('/path/to/db.db')->resultset('users');

=cut

sub sqlite {
    my ($class, $file, $args) = @_;

    $args = {}
        if ! $args;
    # we don't care about making sure the file exists
    # because with sqlite we can just create a new one!
    my $dbh = DBI->connect(
        "dbi:SQLite:$file",
        $args,
    );

    my $dbx = { dbh => $dbh, schema => $class };
    bless $dbx, 'DBIx::Inline::Schema';
}

=head2 config

Sets the location of the configuration (file with the models. The Schema models.. not girls).
This allows you to have the file anywhere on your system and you can rename it to anything.

    # /var/schema/myschemas.yml
    Foo:
      connect: 'SQLite:/var/db/mydb.db'
    
    # /scripts/db.pl
    package main;

    use base 'DBIx::Inline';

    main->config ('/var/schema/myschemas.yml');
    my $schema = main->model('Foo');

You can even chain C<config> to C<model> if you want.

    my $schema = main->config('/var/schema/myschemas.yml')->model('Foo');


=cut

sub config {
    use vars qw/$global/;
    my ($class, $file) = @_;
    
    if ($file) { $global->{config} = $file; return $class }
    else { return $class; }#$global->{config}||0; }
}

=head1 AUTHOR

Brad Haywood <brad@geeksware.net>

=head1 LICENSE

You may distribute this code under the same terms as Perl itself.

=cut

1;
